File: [104] Maximum Depth of Binary Tree/104.maximum-depth-of-binary-tree.java
/*
 * @lc app=leetcode id=104 lang=java
 *
 * [104] Maximum Depth of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        // DFS
        if (root == null) return 0;
        int res = 0;
        Stack<TreeNode> st = new Stack<>();
        Stack<Integer> height = new Stack<>();
        st.push(root);
        height.push(1);
        while (!st.empty()) {
            int depth = height.pop();
            res = Math.max(res, depth);
            root = st.pop();
            if (root.right != null) {
                st.push(root.right);
                height.push(depth+1);
            } 
            if (root.left != null) {
                st.push(root.left);
                height.push(depth+1);
            }
        }
        return res;
    }
}
// @lc code=end

File: [104] Maximum Depth of Binary Tree/104.maximum-depth-of-binary-tree.cpp
/*
 * @lc app=leetcode id=104 lang=cpp
 *
 * [104] Maximum Depth of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        // BFS
        if (root == nullptr) return 0;
        queue<TreeNode*> q;
        int res = 0;
        q.push(root);
        while (!q.empty()) {
            int n = q.size();
            while (n > 0) {
                n--;
                root = q.front();
                q.pop();
                if (root->left != nullptr) {
                    q.push(root->left);
                }
                if (root->right != nullptr) {
                    q.push(root->right);
                }
            }
            res++;
        }
        return res;
    }
};
// @lc code=end

